fe/src/main/java/org/apache/doris/rewrite/FoldConstantsRule.java

```diff
@@ -0,0 +1,233 @@
+        // of the Expr tree. Assumes the bottom-up application of this rule. Constant
+        // children should have been folded at this point.
+        for (Expr child : expr.getChildren()) {
+            if (!child.isLiteral() && !(child instanceof CastExpr)) {
+                return expr;
+            }
+        }
+
+        if (expr.isLiteral() || !expr.isConstant()) {
+            return expr;
+        }
+
+        // Do not constant fold cast(null as dataType) because we cannot preserve the
+        // cast-to-types and that can lead to query failures, e.g., CTAS
+        if (expr instanceof CastExpr) {
+            CastExpr castExpr = (CastExpr) expr;
+            if (castExpr.getChild(0) instanceof NullLiteral) {
+                return expr;
+            }
+        }
+        // Analyze constant exprs, if necessary. Note that the 'expr' may become non-constant
+        // after analysis (e.g., aggregate functions).
+        if (!expr.isAnalyzed()) {
+            expr.analyze(analyzer);
+            if (!expr.isConstant()) {
+                return expr;
+            }
+        }
+        return simplify(expr);
+    }
+
+    private Expr simplify(Expr constExpr) throws AnalysisException {
+        if (constExpr instanceof ArithmeticExpr
+                || constExpr instanceof FunctionCallExpr
+                || constExpr instanceof CastExpr) {
+            Function fn = constExpr.getFn();
+            List<ScalarType> argTypes = new ArrayList<>();
+            for (Type type : fn.getArgs()) {
+                argTypes.add((ScalarType) type);
+            }
+            FEFunctionSignature signature = new FEFunctionSignature(fn.functionName(),
+                    argTypes.toArray(new ScalarType[argTypes.size()]), (ScalarType) fn.getReturnType());
+            FEFunctionInvoker invoker = getFunction(signature);
+            if (invoker != null) {
+                try {
+                    return invoker.invoke(constExpr.getChildrenWithoutCast());
+                } catch (AnalysisException e) {
+                    return constExpr;
+                }
+            }
+        }
+        return constExpr;
+    }
+
+    private FEFunctionInvoker getFunction(FEFunctionSignature signature) {
+        Collection<FEFunctionInvoker> functionInvokers = functions.get(signature.getName());
+
+        if (functionInvokers == null) {
+            return null;
+        }
+
+        for (FEFunctionInvoker invoker : functionInvokers) {
+            if (!invoker.getSignature().returnType.equals(signature.getReturnType())) {
+                continue;
+            }
+
+            ScalarType[] argTypes1 = invoker.getSignature().getArgTypes();
+            ScalarType[] argTypes2 = signature.getArgTypes();
+
+            if (!Arrays.equals(argTypes1, argTypes2)) {
+                continue;
+            }
+
+            return invoker;
+        }
+        return null;
+    }
+
+    private void registerFunctions() {
```

Comment:
This should be synchronized? 
Now, there may concurrent calling

Reply:
Yes.  Fixed.